ocaml/testsuite/tests/typing-layouts-or-null/reexport.ml

```diff
@@ -0,0 +1,234 @@
+       But the kind of 'a Or_null.t must be a subkind of value
+         because of the definition of t at line 2, characters 2-45.
+|}]
+
+(* Incorrect annotations still cause errors. *)
+
+type 'a t : value = 'a or_null [@@or_null_reexport]
+
+[%%expect{|
+Line 1, characters 0-51:
+1 | type 'a t : value = 'a or_null [@@or_null_reexport]
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type 'a or_null is value_or_null
+         because it is the primitive value_or_null type or_null.
+       But the kind of type 'a or_null must be a subkind of value
+         because of the definition of t at line 1, characters 0-51.
+|}]
+
+type 'a t : float64 = 'a or_null [@@or_null_reexport]
+
+[%%expect{|
+Line 1, characters 0-53:
+1 | type 'a t : float64 = 'a or_null [@@or_null_reexport]
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type 'a or_null is value
+         because it is the primitive value_or_null type or_null.
+       But the layout of type 'a or_null must be a sublayout of float64
+         because of the definition of t at line 1, characters 0-53.
+|}]
+
+type ('a : float64) t = 'a or_null [@@or_null_reexport]
+
+[%%expect{|
+Line 1, characters 24-26:
+1 | type ('a : float64) t = 'a or_null [@@or_null_reexport]
+                            ^^
+Error: This type ('a : value) should be an instance of type ('a0 : float64)
+       The layout of 'a is float64
+         because of the annotation on 'a in the declaration of the type t.
+       But the layout of 'a must overlap with value
+         because the type argument of or_null has layout value.
+|}]
+
+(* The behavior here is correct but confusing: ['a] is restricted to [value],
+   which is a subjkind of [value_or_null]. *)
+
+module Or_null = struct
+  type ('a : value_or_null) t = 'a or_null [@@or_null_reexport]
+end
+let fail = Or_null.This (Or_null.This 5)
+
+[%%expect{|
+module Or_null : sig type 'a t = 'a or_null = Null | This of 'a end
+Line 4, characters 24-40:
+4 | let fail = Or_null.This (Or_null.This 5)
+                            ^^^^^^^^^^^^^^^^
+Error: This expression has type 'a Or_null.t = 'a or_null
+       but an expression was expected of type ('b : value)
+       The kind of 'a Or_null.t is value_or_null
+         because it is the primitive value_or_null type or_null.
+       But the kind of 'a Or_null.t must be a subkind of value
+         because of the definition of t at line 2, characters 2-63.
+|}]
+
+(* This fails, just as [type t = int option = None | Some of int] would. *)
+
+type t = int or_null [@@or_null_reexport]
+
+[%%expect{|
+Line 1, characters 0-41:
+1 | type t = int or_null [@@or_null_reexport]
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This variant or record definition does not match that of type
+         int or_null
+       They have different arities.
+|}]
+
+(* [@@or_null_reexport] requires a manifest. *)
+
+type 'a t [@@or_null_reexport]
```

Comment:
It would be cool to accept this, for the following idiom:

```ocaml
module M : sig
  type 'a t [@@or_null_reexport]
end = struct
  type 'a t = 'a or_null [@@or_null_reexport]
end
```

Now, `M.t` behaves just like `or_null` but is still distinct. Which might be useful. That said, you've described this as hard, so we shouldn't. So all of this is to say: leave a comment here saying this would be nice, but we'll wait for an actual use-case.